### ISSUE TITLE
resource returned with 'spn: {id}'

### ISSUE THREAD

**Issue by nazarsh** [AzureAD/azure-activedirectory-library-for-nodejs#128]
Hello and thank you for providing this library!

having passed a resource of `00000002-0000-0000-c000-000000000000`

I get back a token which has 
`resource: 'spn:00000002-0000-0000-c000-000000000000'`
          ^ note the spn

This seems to differ from the Angular version of the library which returns the resource without the spn. Is there a parameter I can pass to suppress this behaviour, or is that something that is desired/expected?

Thanks!

**Comment by admeister** [comment]
+1

**Comment by danieljin** [comment]
BUMP! Manually did the change of setting api version to None and it stripped the spn:.
https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/dev/lib/oauth2client.js#L90


Can we get the same change as the python library?

**Comment by rayluo** [comment]
Thanks for everyone who participated this conversation thread. The different behavior of that resource value is indeed controlled by the presence or absence of the "api-version=1.0". However we do not clearly know how many customers out there are unknowningly relying on the current "api-version=1.0" behavior, so we do not want to simply remove it and potentially break existing customer's code. The change we did for ADAL Python library, i.e. exposing the api-version as an optional parameter, is a temporary solution. We will revisit this later. Meanwhile, feel free to adapt the PR from @dstrimble as a workaround for now. (Thanks @dstrimble !)

**Comment by ashishtechie** [comment]
I am still seeing this issue, actually i saw this issue when i was trying to see that caching on the library is not working. The reason why its not working is because when it compares the resource it finds the 'spn:' in the beginning and basically creates a new token for the same request. Can anybody help on that, if i am missing anything?

**Comment by jaimemm** [comment]
Hi,

I am also experiencing the same behaviour, which causes a "IDX10214: Audience validation failed" error when invoking my API App URL.
Is it there any workaround to get the spn prefix removed from the audiences?

Regards

**Comment by myunggeun-song** [comment]
Hi,

Would you let us know any progress or decision for this problem? We are using the HACK but it should not be a permanent option. Please take care of this issue. 

Thanks!

**Comment by jaimemm** [comment]
At least please reply if you're accepting pull requests, so we can fix it.

Regards

**Comment by rayluo** [comment]
Thanks for your interest in helping. I've updated my [earlier comment](https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/issues/128#issuecomment-287457624) to include the way we introduced a workaround for this in ADAL Python. A pull request in ADAL Nodejs will need to follow similar approach.

**Comment by hashtagchris** [comment]
Regarding customers relying on the current "api-version=1.0" behavior: Please consider introducing a new major version (e.g. 1.0.0) of the adal-node npm package so you can break backward compatibility per https://docs.npmjs.com/getting-started/semantic-versioning

**Comment by rayluo** [comment]
Thanks @hashtagchris ! That is the plan. We will try to include more shiny new features in the 1.0 version too.

**Comment by salvoravida** [comment]
please merge : 

https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/pull/186

**Comment by scott-lin** [comment]
The "api-version=1.0" workaround isn't possible in TypeScript if I'm not mistaken since AuthenticationContext constructor doesn't expose a way to set the value.

https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/606fa2a0b365f75848305033b0d964476820edef/lib/adal.d.ts#L277

@rayluo, any suggestions?

**Comment by scott-lin** [comment]
I plan on using the aforementioned [workaround](https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/pull/149) for now, but is there any update on the proper fix?

**Comment by rayluo** [comment]
Sorry we don't currently have an update for this one. I'm looping our PM @navyasric into this conversation in order to keep track of the user voice.

**Comment by polys** [comment]
This has been open for ages and still no proper fix? Is this still the official library or should we be using something else instead?

**Comment by gmoshe27** [comment]
It's a shame that there isn't yet a work-around, as it would save many users a lot of time. In the interim, if you are looking for a way to get a client credential, that's auth with client id and client secret, then here is what I figured out:

Microsoft documents the http call for authenticating a service to a service:
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols-oauth-client-creds#first-case-access-token-request-with-a-shared-secret

I found that if you remove the `/v2.0/` path in the request url, then you can get a client secret without needing to specify scopes. This is how the .Net library handles it.

You only need to make a single request to get the bearer token with client credentials:
- tenant - this is your organizations tenant company.onmicrosoft.com
- client_id - this is also known as the Application ID in azure portal
- client_secret - this is the secret key that you create in your AD app in the Keys/Passwords section
- grant_type - set this to `client_credentials`
- resource - this is the GUID of the resource you want, in my case I have an app I own connecting to an api that I own, so the `resource` and the `client_id` are the same.

```
POST: https://login.microsoftonline.com/{tenant}/oauth2/token
Content-Type: application/x-www-form-urlencoded

client_id=535fb089-9ff3-47b6-9bfb-4f1264799865
&resource=535fb089-9ff3-47b6-9bfb-4f1264799865
&client_secret={secret that you create for your app}
&grant_type=client_credentials
```

You can test it in postman and make your own wrappers around the response in node. For handling token expiration and refreshes, I found that the `simple-oauth2` package had a good description of the gotchas and how to handle them. With the approach above, the response does seem to come back with `expires_in` which is not as exact as `expires_at`.

https://github.com/lelylan/simple-oauth2#helpers

Good luck

**Comment by ryanzjohnson-ms** [comment]
I am having the same issue but with Client Token validation. Any update on this?

**Comment by hashtagchris** [comment]
Based on #149 I removed the api-version parameter and published `@hashtagchris/adal-node-no-api-version` to the npm registry, if people want to use that package until this issue is fixed.

**Comment by dylanjmorgan** [comment]
This has been going on for 3 years!  I stumbed on the issue writing an interface to AAS in Node (to run on linux) where Microsoft expects the resource to be https://*.asazure.windows.net.

@hashtagchris your repo works, where the official one is broken, apparently for 3 years.  Thanks!

**Comment by rayluo** [comment]
Thanks for all your patience. This is now fixed in [release 0.2.0](https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/releases/tag/0.2.0) after we merged #186. You will still need to manually opt-in for the new behavior.
CC @DarylThayil
